File: p2/code/depthFromStereo.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
from scipy.spatial.distance import cdist

def depthFromStereo(img1, img2, ws):
    img1 = np.dot(img1[...,:3], [0.2989, 0.5870, 0.1140])
    img2 = np.dot(img2[...,:3], [0.2989, 0.5870, 0.1140])

    depth_map = np.ones((img1.shape[0], img1.shape[1]))

    img1 = np.pad(img1, ws, mode='constant')
    img2 = np.pad(img2, ws, mode='constant')
    # print(img1.shape, img2.shape)
    disparity_mat = np.zeros((img1.shape[0], img1.shape[1]))

    for i in range(ws,img1.shape[0]):
        for j in range(ws, img2.shape[1]):
            min_ssd = 100000
            for k in range(ws, img2.shape[1]):

                try:
                    # ssd = np.sum((np.subtract(img1[i-ws:i+ws,j-ws:j+ws], img2[i-ws:i+ws,k-ws:k+ws]))**2)
                    ssd = cdist(img1[i-ws:i+ws, j-ws:j+ws], img2[i-ws:i+ws, k-ws:k+ws], 'sqeuclidean')
                    if(ssd < min_ssd):
                        min_ssd = ssd
                        x = np.absolute(k-j)
                except:
                    break

            disparity_mat[i - ws, j - ws] = x

        print(i, j)

    return np.divide(1, disparity_mat)

File: p2/code/depthFromStereo.m
function depth = depthFromStereo(img1, img2, ws)
% Implement this
depth = ones(size(img1,1), size(img1,2));
img1 = im2double(rgb2gray(img1));
img2 = im2double(rgb2gray(img2));

disparityMap = disparitySGM(frameLeftGray, frameRightGray);
figure;
imshow(disparityMap, [0, 64]);
title('Disparity Map');
colormap jet
colorbar

[nRows,nCols] = size(depth);

% Pad both images with 0
img1 = padarray(img1,[ws ws],0,'both');
img2 = padarray(img2,[ws ws],0,'both');

disp_mat = zeros(size(img1,1), size(img1,2));

% Calculate the displacement for each pixel
for i = ws+1:nRows+ws
    for j = ws+1:nCols+ws
        minSSD = inf;
        disp_x = 0;
        patch1 = img1(i-ws:i+ws,j-ws:j+ws);
        for k = ws+1:nCols+ws
                patch2 = img2(i-ws:i+ws,k-ws:k+ws);
%               Using SSD to find the correspondences
                ssd = sum(sum((patch1-patch2).^2));
                if ssd < minSSD
                    minSSD = ssd;
                    disp_x = abs(k-j);
                end
        end
    disp_mat(i-ws,j-ws) = disp_x;
    end
end
% Calculate Depth using Disparity since b.f=1
depth = 1./disp_mat;
end